service/src/main/java/com/epam/esm/gifts/impl/GiftCertificateServiceImpl.java

```diff
@@ -84,43 +88,59 @@ public GiftCertificateDto findById(Long id) {
+        if (!validator.isAttributeDtoValid(attributeDto)) {
+            throw new SystemException(INVALID_ATTRIBUTE_LIST);
+        }
+        setDefaultParamsIfAbsent(attributeDto);
+        Pageable sortedPageable = buildPageableSort(attributeDto.getSortingFieldList(), attributeDto.getOrderSort(), pageable);
+        Page<GiftCertificate> certificatePage = Objects.nonNull(attributeDto.getTagNameList())
+                ? giftCertificateRepository.findByAttributes(attributeDto.getTagNameList()
+                , attributeDto.getTagNameList().size(), attributeDto.getSearchPart(), sortedPageable)
+                : giftCertificateRepository.findByAttributes(attributeDto.getSearchPart(), sortedPageable);
+        if (!validator.isPageExists(pageable, certificatePage.getTotalElements())) {
             throw new SystemException(NON_EXISTENT_PAGE);
         }
-        int offset = calculateOffset(pageable);
-        List<GiftCertificate> certificateList = giftCertificateDao.findByAttributes(attribute, offset, pageable.getSize());
-        List<GiftCertificateDto> certificateDtoList = certificateList.stream().map(giftCertificateConverter::giftCertificateToDto).toList();
-        return new CustomPage<>(certificateDtoList, pageable, totalCertificateNumber);
+        return new CustomPage<>(certificatePage.getContent(), certificatePage.getPageable(), certificatePage.getTotalElements())
+                .map(giftCertificateConverter::giftCertificateToDto);
+    }
+
+    private void setDefaultParamsIfAbsent(GiftCertificateAttributeDto attributeDto) {
+        if (Objects.isNull(attributeDto.getSearchPart())) {
+            attributeDto.setSearchPart("");
+        }
+    }
+
+    private Pageable buildPageableSort(List<String> sortingFieldList, String orderSort, Pageable pageable) {
+        Sort.Direction direction = Objects.nonNull(orderSort)
+                ? Sort.Direction.fromString(orderSort)
+                : Sort.Direction.ASC;
+        return CollectionUtils.isEmpty(sortingFieldList)
+                ? PageRequest.of(pageable.getPageNumber(), pageable.getPageSize(), Sort.by(direction, "id"))
+                : PageRequest.of(pageable.getPageNumber(), pageable.getPageSize()
+                , by(direction, sortingFieldList.toArray(String[]::new)));
     }
 
     @Override
     @Transactional
     public void delete(Long id) {
-        Optional<GiftCertificate> optionalGiftCertificate = giftCertificateDao.findById(id);
-        if (optionalGiftCertificate.isPresent()) {
-            if (giftCertificateDao.isGiftCertificateUsedInOrders(id)) {
-                giftCertificateDao.delete(optionalGiftCertificate.get());
-            } else {
-                throw new SystemException(USED_ENTITY);
-            }
+        Optional<GiftCertificate> optionalGiftCertificate = giftCertificateRepository.findById(id);
+        if (orderRepository.findFirstByCertificateListId(id).isPresent()) {
```

Comment:
Yes, you are right, since Java 9, a method has really appeared
it will looks like 
orderRepository.findFirstByCertificateListId(id).ifPresentOrElse((a)->giftCertificateRepository.delete(optionalGiftCertificate.get()),
                ()->{throw new SystemException(NON_EXISTENT_ENTITY);});
but there were some problems with the lambda